### PR TITLE
Fix EntityDataMap serialization to only write present flag groups

### LINKED ISSUE #219: Incorrect Entity Flags translation
## Description:
Currently, entity flag translation is inaccurate. Evidences pointing to this are mentioned in #49 and by testing a PMMP shields implementation.

Instead of lowering the shield properly, it would give you the levitation effect and the underwater breathing timer + also flicker the screen.

## Screenshots / videos:

https://user-images.githubusercontent.com/17762324/187044541-cb70af11-9df5-486e-8f0a-e43ced1e5f63.mp4

### PR DESCRIPTION
**Pull request by irrld**
Fixes https://github.com/WaterdogPE/WaterdogPE/issues/219#issuecomment-3382829653

When EntityDataMap is written, it writes both FLAGS and FLAGS_2 if either of them exists. However, if only one flag was present during reading (for example, if FLAGS_2 was sent without FLAGS), the missing flag would not be read but would still be written as zero during serialization. This creates a read/write mismatch that can be tested in ProxyPass by sending the read packet instead of the read buffer, forcing it to re-encode when sending to the client. This issue even affects BDS and causes client to see the missing flag set to zero, which clears it unintentionally.

This PR changes how EntityDataMap is serialized and what FlagTransformer returns when it finds no matching flags. FlagTransformer will now return null if there are no matching flags for the given part. When serializing EntityDataMap, it first collects the non-null serialized values, then writes them to the buffer.

### CODE REVIEW

File: bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java [hunk 1]
```diff
@@ -24,14 +24,18 @@ public Long serialize(BedrockCodecHelper helper, EntityDataMap map, EnumSet<Enti
...
                 value |= 1L << (flagIndex & 0x3f);
+                hasValue = true;
             }
         }
-
-        return value;
+        if (hasValue) {
+            return value;
+        }
+        return null;
```
Comment (SupremeMortal) [inline bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java]: This logic looks flawed. Say we have a flag that is currently set to `true` on the client and we want to set that to `false`, this code looks as if it won't get sent at all.

**Review by SupremeMortal — CHANGES_REQUESTED** [review]
I see where you're going with these changes, but I think it's going to require more extensive changes to achieve what you want.

File: bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java (hunk 1 above)
Comment (SupremeMortal) [reply bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java]: Really, for this to work the same as vanilla, we would need to switch the flags from an EnumSet to an EnumMap to allow `true`, `false` or `null` as values for a flag.

Comment (irrld) [reply bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java]: Hmm I see what you mean, yeah it is flawed. 

I will take the PR to draft and rethink our approach, storing the flags in one enum and having dynamic indexes for each flag makes this a difficult problem to solve.

Comment (irrld) [reply bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java]: > Really, for this to work the same as vanilla, we would need to switch the flags from an EnumSet to an EnumMap to allow `true`, `false` or `null` as values for a flag.

Do you think we should do that or find another solution?

Comment (SupremeMortal) [reply bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java]: I think that's the most effective way to solve the problem. It's going to be an API breakage, but I think that's not much of a problem since we're still not in released state, and I don't believe it's much hassle for downstream projects to fix.

Comment (sircurse) [reply bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java]: you don’t need to switch to an `EnumMap<Flag, Boolean?>` to get vanilla-correct behavior. the protocol itself already gives you the tri-state you want at the **field** level, not per-flag:

* **field absent** -> “no change”
* **field present with mask bit = 1** -> that bit is **true**
* **field present with mask bit = 0** -> that bit is **false**

so you only need to decide **when to emit the lane**, not store `true/false/null` per individual flag. keeping `EnumSet<Flag>` for the **true** flags is fine

Comment (SupremeMortal) [reply bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java]: So if I'm a consumer of the protocol library and I want to set only the flag at index 138 to `false` in the second bitflags, how would you achieve this without sending the first flags, but still sending the current flags as an EnumSet? Sure we can expose the flags downstream as something nicer than an EnumMap, but we still need to know if the consumer wants to set something to true, false, or not at all to be able to determine if we should  send the certain bitflags.

File: bedrock-codec/src/test/java/org/cloudburstmc/protocol/bedrock/test/FlagSerializationTest.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+
+public class FlagSerializationTest {
+    private static final BedrockCodecHelper CODEC_HELPER = Bedrock_v776.CODEC.createHelper();
+
+    private static final TypeMap<EntityFlag> ENTITY_FLAGS;
+
+    static {
+        try {
+            Field field = Bedrock_v776.class.getDeclaredField("ENTITY_FLAGS");
```
Comment (irrld) [inline bedrock-codec/src/test/java/org/cloudburstmc/protocol/bedrock/test/FlagSerializationTest.java:25]: I'm not very happy of using reflection here to get the flags but it would require a lot more change to get it otherwise.

**Comment by irrld** [comment]
Changed EnumSet to EnumMap and added tests, let me know if you can think of more edge cases we can test.

File: bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java (hunk 1 above)
Comment (sircurse) [reply bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java]: 138 is outside lane 2 in current Bedrock—lane 2, it is 64..127, but let me explain

1. Build the **target** masks from the `EnumSet`:
   * `lo = packLane(flags, 0..63)`
   * `hi = packLane(flags, 64..127)`
2. Compare to **last sent** masks (cache per entity+client): `lastLo`, `lastHi`.
3. Emit deltas **only for lanes that changed**:
   * If `hi != lastHi` ⇒ write `FLAGS_2 = hi` (even if `hi == 0`, so the client clears).
   * If `lo != lastLo` ⇒ write `FLAGS = lo` (only if changed).
4. Update cache.

that lets a consumers “set one flag to false” by just **removing** it from the EnumSet and calling encode; the library figures out the lane mask changed and emits **only `FLAGS_2`** with the cleared bit. No need to send `FLAGS`. No need for a per-flag tri-state.

to fix it you could do something like this on Transformers, this is what we did on our code to fix that:
```
private static long packEntityFlags(EnumSet<EntityFlag> set, int base) {
	long bits = 0L;
	if (set == null || set.isEmpty()) return bits;
	for (EntityFlag f : set) {
		int id = f.getValue();
		if (id >= base && id < base + 64) {
			bits |= (1L << (id & 0x3F));
		}
	}
	return bits;
}
```

Comment (irrld) [reply bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java]: This is a protocol lib, we don't store and cache it, there is no logic that allows such thing. That cannot be applied to here.

Comment (irrld) [reply bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java]: You are looking this from the wrong perspective, protocol lib allows reading a packet from a version and writing to another, flags indexes can and do change. We cannot just store numbers.

Comment (irrld) [reply bedrock-codec/src/main/java/org/cloudburstmc/protocol/bedrock/transformer/FlagTransformer.java]: Packets are abstract from the version logic, and because protocol lib considers both FLAGS and FLAGS_2 as the same, you have to have per-flag tri-state to know if given flag exists in one of the groups.

File: bedrock-codec/src/test/java/org/cloudburstmc/protocol/bedrock/test/FlagSerializationTest.java (hunk 2 above)
Comment (Alemiz112) [reply bedrock-codec/src/test/java/org/cloudburstmc/protocol/bedrock/test/FlagSerializationTest.java:25]: should this be under adventure module?

Comment (irrld) [reply bedrock-codec/src/test/java/org/cloudburstmc/protocol/bedrock/test/FlagSerializationTest.java:25]: Uh not really, moved it to codec.

**Comment by irrld** [comment]
Hi, what is the status for this PR?

**Review by SupremeMortal — APPROVED** [review]
